TypeError: ɵgetOrCreateAngularServerApp is not a function in 19.1.x after fix angular-compiler error

### Command

serve

### Is this a regression?

- [x] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

_No response_

### Description

Throw error message when hot reload in 19.1.x, this issue still exists in v19.1.4.

![Image](https://github.com/user-attachments/assets/f9d26f96-4d83-43f9-9f5f-4a5b9b9ae144)

![Image](https://github.com/user-attachments/assets/433ec5ab-add0-4bff-a7e5-ab874f945f92)

https://github.com/angular/angular-cli/issues/29379#issuecomment-2604998446


### Minimal Reproduction

1. Edit the project to throw angular-compiler error.
eg. 
`X [ERROR] NG5002: Unexpected character "}" [plugin angular-compiler]`
or 
`X [ERROR] TS1128: Declaration or statement expected. [plugin angular-compiler]`
2. Fix the code.
3. Observe the error in the terminal and the browser.

### Exception or Error

```text
TypeError: ɵgetOrCreateAngularServerApp is not a function
    at _AngularAppEngine.getAngularServerAppForRequest (C:\Project\rosy-cart\.angular\cache\19.1.4\rosy-cart\vite\deps_ssr\chunk-VBOT64B7.js:11367:23)
    at async _AngularAppEngine.handle (C:\Project\rosy-cart\.angular\cache\19.1.4\rosy-cart\vite\deps_ssr\chunk-VBOT64B7.js:11306:23)
```

### Your Environment

```text
Angular CLI: 19.1.4
Node: 20.18.1
Package Manager: pnpm 10.0.0
OS: win32 x64

Angular: 19.1.3
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1901.4
@angular-devkit/build-angular   19.1.4
@angular-devkit/core            19.1.4
@angular-devkit/schematics      19.1.4
@angular/cli                    19.1.4
@angular/ssr                    19.1.4
@schematics/angular             19.1.4
rxjs                            7.8.1
typescript                      5.5.4
```

### Anything else relevant?

_No response_

--- a/packages/angular/build/src/builders/dev-server/vite-server.ts
+++ b/packages/angular/build/src/builders/dev-server/vite-server.ts
@@ -501,25 +501,15 @@ async function invalidateUpdatedFiles(
   }
 
   // Invalidate any updated files
-  let destroyAngularServerAppCalled = false;
+  let serverApplicationChanged = false;
   for (const [file, record] of generatedFiles) {
     if (!record.updated) {
       continue;
     }
+
     record.updated = false;
-
-    if (record.type === BuildOutputFileType.ServerApplication && !destroyAngularServerAppCalled) {
-      // Clear the server app cache
-      // This must be done before module invalidation.
-      const { ɵdestroyAngularServerApp } = (await server.ssrLoadModule('/main.server.mjs')) as {
-        ɵdestroyAngularServerApp: typeof destroyAngularServerApp;
-      };
-
-      ɵdestroyAngularServerApp();
-      destroyAngularServerAppCalled = true;
-    }
-
     updatedFiles.push(file);
+    serverApplicationChanged ||= record.type === BuildOutputFileType.ServerApplication;
 
     const updatedModules = server.moduleGraph.getModulesByFile(
       normalizePath(join(server.config.root, file)),
@@ -527,9 +517,14 @@ async function invalidateUpdatedFiles(
     updatedModules?.forEach((m) => server.moduleGraph.invalidateModule(m));
   }
 
-  if (destroyAngularServerAppCalled) {
-    // Trigger module evaluation before reload to initiate dependency optimization.
-    await server.ssrLoadModule('/main.server.mjs');
+  if (serverApplicationChanged) {
+    // Clear the server app cache and
+    // trigger module evaluation before reload to initiate dependency optimization.
+    const { ɵdestroyAngularServerApp } = (await server.ssrLoadModule('/main.server.mjs')) as {
+      ɵdestroyAngularServerApp: typeof destroyAngularServerApp;
+    };
+
+    ɵdestroyAngularServerApp();
   }
 
   return updatedFiles;

--- a/packages/angular/build/src/tools/vite/middlewares/ssr-middleware.ts
+++ b/packages/angular/build/src/tools/vite/middlewares/ssr-middleware.ts
@@ -40,6 +40,11 @@ export function createAngularSsrInternalMiddleware(
       const { writeResponseToNodeResponse, createWebRequestFromNodeRequest } =
         await loadEsmModule<typeof import('@angular/ssr/node')>('@angular/ssr/node');
 
+      // The following is necessary because accessing the module after invalidation may result in an empty module,
+      // which can trigger a `TypeError: ɵgetOrCreateAngularServerApp is not a function` error.
+      // TODO: look into why.
+      await server.ssrLoadModule('/main.server.mjs');
+
       const { ɵgetOrCreateAngularServerApp } = (await server.ssrLoadModule('/main.server.mjs')) as {
         ɵgetOrCreateAngularServerApp: typeof getOrCreateAngularServerApp;
       };

